README.md
ServerUsage - README
====================

+ Name: ServerUsage

+ Version: 6.3.8

+ Release date: 2012-11-29

+ Author: Nicola Asuni

+ Copyright (2012-2012):

> > Fubra Limited  
> > Manor Coach House  
> > Church Hill  
> > Aldershot  
> > Hampshire  
> > GU12 4RQ  
> > <http://www.fubra.com>  
> > <support@fubra.com>  


SOFTWARE LICENSE:
-----------------

This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

See LICENSE.TXT file for more information.


DESCRIPTION:
------------

The ServerUsage system is designed to collect and process statistic information from computers running a GNU-Linux Operating System.

This project is composed by:


## ServerUsage-Server ##

The ServerUsage-Server program listen on a TCP port for incoming log data (from ServerUsage-Client) and store them on a SQLite table. An script (serverusage_dbagg.sh) is executed periodically to aggregate data on another table and delete obsolete data.

The serverusage_api.php script can be remotely used to extract formatted data from the database or display graphs.


## ServerUsage-Client ##

The ServerUsage-Client program is composed by two main sections: the SystemTap (http://sourceware.org/systemtap) serverusage_client.ko kernel module to collect and output usage statistics of the machine where it is installed, and the serverusage_tcpsender.bin to send the output to a log server via TCP.


## ServerUsage-Client-MDB ##

The ServerUsage-Client-MDB program is used to collect user statistics form a MariaDB database.



## GENERAL USAGE SCHEMA ##
	
	(ServerUsage-Client) - - TCP CONNECTION - -> (ServerUsage-Server)

	([serverusage_client.ko]==>[serverusage_tcpsender.bin]) - - TCP CONNECTION - -> ([serverusage_tcpreceiver]==>[SQLite log_raw table (raw data)])

	[SQLite log_raw table]==>[serverusage_dbagg.sh]==>[SQLite log_agg_hst table (aggregated data)]

	[SQLite log_agg_hst table]==>[serverusage_api.php]==>[Data with selected format or SVG graph]

## API ##

The serverusage_api.php script can be remotely used to extract formatted data from the database or display graphs.

### PARAMETERS: ###

	from: (integer) starting timestamp in seconds since EPOCH;
	to: (integer) starting timestamp in seconds since EPOCH - this value can be at maximum equal to (current_time - (2 * DB_AGGREGATION_DELAY)) that is also the default value;
	metric: (not available with svg mode) type of info to extract; Possible values are: 'uid', 'ip', 'uip', 'grp', 'glb', 'all'. The return values for each metric are:
		uid : user_id, cpu_ticks;
		uidt : user_id, cpu_ticks, minimum start time, maximum end time;
		ip  : ip, net_in, net_out;
		ipt  : ip, net_in, net_out, minimum start time, maximum end time;
		uip : user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out;
		uipt : user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out, minimum start time, maximum end time;
		grp : start_time, end_time, user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out;
		glb : (default for SVG mode) lah_start_time, lah_end_time, lah_cpu_ticks, lah_io_read, lah_io_write, lah_netin, lah_netout;
		all : start_time, end_time, process, user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out.
	uid : (integer) if set, filter result for the requested user ID;
	ip : (IP address) if set, filter result for the requested IP address.
	mode: output format ('json' = JSON, 'csv' = CSV TAB-Separated Values, 'psa' = base64 encoded PHP Serialized array, 'svg' = SVG).

	ADDITIONAL PARAMETERS FOR SVG MODE:

	width: (integer) optional width for SVG output (default 1024; minimum 50).
	height: (integer) optional height for SVG output; will be rounded to a multiple of 5 (default 750, minimum 50).
	scale: linear = vertical linear scale (default), log = vertical logarithmic scale.
	bgcol: type of background color: 'dark' or 'light' (default). Note: the SVG background is always transparent.
	gtype: sequence of number representing the graphs to display: 1 = CPU TICKS, 2 = IO READ, 3 = IO WRITE, 4 = NET IN, 5 = NET OUT. Default: 12345.

### USAGE EXAMPLES: ###

	JSON:
		serverusage_api.php?from=1332769800&to=1332845100&metric=uid&mode=json
		serverusage_api.php?from=1332769800&metric=uidt&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=ip&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=ipt&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=uip&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=uipt&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&uid=320&mode=json

	CSV:
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&uid=320&mode=csv

	BASE64 ENCODED PHP SERIALIZED ARRAY:
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&uid=320&mode=psa

	SVG:
		serverusage_api.php?from=1332769800&to=1332845100&mode=svg&width=1024&height=750&scale=log
		serverusage_api.php?from=1333532663&to=1333627917&mode=svg&scale=log&bgcol=light&gtype=12345
		serverusage_api.php?from=1333532663&to=1333627917&mode=svg&scale=linear&bgcol=light&gtype=15
		serverusage_api.php?from=1333532663&to=1333627917&mode=svg&scale=log&bgcol=light&gtype=5

### OUTPUT: ###

	The output format can be JSON (JavaScript Object Notation), CSV (tab-separated text values), Base64 encoded serialized array or SVG (Scalable Vector Graphics).

### ADDITIONAL NOTES: ###

	The reference time for all servers should be the standard UTC.
	The latest available time on the ServerUsage-Server aggregated table is always in the past by the value specified by DB_AGGREGATION_DELAY constant (by default 5 minutes).
	A valid time interval, for example, can be calculated as follows:

		polling_interval = 900; // 15 minutes * 60 seconds; must be equal or greater than DB_AGGREGATION_DELAY.
		delay_time = 600; // 10 minutes * 60 seconds; must be equal or greater than (2 * DB_AGGREGATION_DELAY).
		end_time = (current_time - delay_time);
		start_time = (end_time - polling_interval);

HOW-TO CREATE ServerUsage RPMs
------------------------------

This is a short hands-on tutorial on creating RPM files for the ServerUsage project.
For an automatic building script for CentOS and the latest RPM packages please check the CatN Repository: https://github.com/fubralimited/CatN-Repo

NOTE: The sever configuration for ServerUsage-Client and ServerUsage-Server may be different, so this process must be executed in different environments.

## DEVELOPMENT ENVIRONMENT ##

To build RPMs we need a set of development tools.

CHANGELOG.TXT
6.3.8 (2013-03-19)
	- Server init script was rewritten to be compatible with other Linux distributions.

6.3.6 (2012-11-29)
	- The serverusa_api.php script was improved with additional controls and default values for to/from parameters.

6.3.5 (2012-11-15)
	- The serverusa_api.php script was improved to include new metric options. Read the documentation on the file source code for more information.

6.3.3 (2012-09-18)
	- Programs are now properly daemonized.
	- Init scripts were fixed.

6.3.2 (2012-09-17)
	- Init scripts were improved.

6.3.1 (2012-09-03)
	- String to number conversion functions were fixed.

6.3.0 (2012-08-29)
	- Debug mode on server was improved to display milliseconds.

6.2.0 (2012-08-21)
	- RPM spec files were changed to avoid overriding the configuration and database files.
	- Init scripts were improved.

6.1.0 (2012-08-16)
	- This versin contains a new module to collect user statistics from MariaDB database.

6.0.0 (2012-08-09)
	- Dual stack support for both IPv4 and IPv6 was added.

5.3.0 (2012-08-08)
	- Documentation was updated.

5.2.0 (2012-07-18)
	- Switch to using contention-free aggregates for the systemtap script.
	- SystemTap skipped probes are now ignored to avoid early program termination.

5.1.0 (2012-07-17)
	- A segmentation fault problems were fixed.

5.0.0 (2012-07-16)
	- Various memory allocation and segmentation fault problems were fixed.

4.9.0 (2012-07-13)
	- A memory allocation problem was fixed.

4.8.0 (2012-06-13)
	- TCP sender was improved to avoid infinite loops.

4.7.0 (2012-06-13)
	- TCP sender and receiver are set to use IPv6 protocol by default.

4.6.0 (2012-06-08)
	- TCP sender and receiver are now able to reuse previously unclosed connections.

4.5.0 (2012-05-22)
	- The program serverusage_tcpreceiver.c was improved to be thread safe.

4.4.0 (2012-05-21)
	- SQLite transactions where fixed to properly lock/unlock the table.

4.3.0 (2012-05-18)
	- serverusage_tcpreceiver.c was fixed.

4.2.0 (2012-05-18)
	- Bug related to nested transactions were fixed.

4.1.0 (2012-05-18)
	- C scripts were analysed and optimised using a memory profiler.

4.0.0 (2012-05-17)
	- This version uses SQLite as back-end database for the server module.
	- Database is now created by makefile script (and preserved if already exist).
	- Init script is now able to install/uninstall the cron job for the database aggregator.
	- The output of the SystemTap module was changed to be TAB-separated instead of comma-separated.

--------------------------------------------------------------------------------
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
----------------------------------------------------------------------------

3.0.4 (2012-05-11)
	- serverusage_tcpsender is now able to store log data on a local filesystem in case of network errors.
	- PATH_TEMPLOG was added into serverusage_client.conf to define the path of local caching.

3.0.3 (2012-05-08)
	- A memory leak was fixed.

3.0.2 (2012-05-08)
	- Fixed minor syntax errors.

3.0.1 (2012-05-08)
	- Minor documentation improvements.

3.0.0 (2012-05-04)
	- This version introduce multithreading support for the TCP receiver.
	- Documentation was fixed.

--------------------------------------------------------------------------------
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
--------------------------------------------------------------------------------

2.0.9 (2012-05-02)

server/serverusage_api.php
<?php
//=============================================================================+
// File name   : serverusage_api.php
// Begin       : 2012-03-12
// Last Update : 2013-01-18
// Version     : 6.3.8
//
// Website     : https://github.com/fubralimited/ServerUsage
//
// Description : Web service to extract and return information
//               from ServerUsage log_agg_hst database table.
//
// Author: Nicola Asuni
//
// (c) Copyright:
//               Fubra Limited
//               Manor Coach House
//               Church Hill
//               Aldershot
//               Hampshire
//               GU12 4RQ
// 							 UK
//               http://www.rackmap.net
//               support@rackmap.net
//
// License:
//    Copyright (C) 2012-2012 Fubra Limited
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU Affero General Public License as
//    published by the Free Software Foundation, either version 3 of the
//    License, or (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU Affero General Public License for more details.
//
//    You should have received a copy of the GNU Affero General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//    See LICENSE.TXT file for more information.
//=============================================================================+

/**
 * @file
 * Web service to extract and return information from ServerUsage log_agg_hst database table.
 * @package com.fubra.serverusage
 * @brief Web service to extract and return information from ServerUsage log_agg_hst database table.
 * @author Nicola Asuni
 * @since 2012-03-12
 */

/*

PARAMETERS:

	from: (integer) starting timestamp in seconds since EPOCH;
	to: (integer) starting timestamp in seconds since EPOCH - this value can be at maximum equal to (current_time - (2 * DB_AGGREGATION_DELAY)) that is also the default value;
	metric: (not available with svg mode) type of info to extract; Possible values are: 'uid', 'ip', 'uip', 'grp', 'glb', 'all'. The return values for each metric are:
		uid : user_id, cpu_ticks;
		uidt : user_id, cpu_ticks, minimum start time, maximum end time;
		ip  : ip, net_in, net_out;
		ipt  : ip, net_in, net_out, minimum start time, maximum end time;
		uip : user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out;
		uipt : user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out, minimum start time, maximum end time;
		grp : start_time, end_time, user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out;
		glb : (default for SVG mode) lah_start_time, lah_end_time, lah_cpu_ticks, lah_io_read, lah_io_write, lah_netin, lah_netout;
		all : start_time, end_time, process, user_id, ip, cpu_ticks, io_read, io_write, net_in, net_out.
	uid : (integer) if set, filter result for the requested user ID;
	ip : (IP address) if set, filter result for the requested IP address.
	mode: output format ('json' = JSON, 'csv' = CSV TAB-Separated Values, 'psa' = base64 encoded PHP Serialized array, 'svg' = SVG).

	ADDITIONAL PARAMETERS FOR SVG MODE:

	width: (integer) optional width for SVG output (default 1024; minimum 50).
	height: (integer) optional height for SVG output; will be rounded to a multiple of 5 (default 750, minimum 50).
	scale: linear = vertical linear scale (default), log = vertical logarithmic scale.
	bgcol: type of background color: 'dark' or 'light' (default). Note: the SVG background is always transparent.
	gtype: sequence of number representing the graphs to display: 1 = CPU TICKS, 2 = IO READ, 3 = IO WRITE, 4 = NET IN, 5 = NET OUT. Default: 12345.

USAGE EXAMPLES:

	JSON:
		serverusage_api.php?from=1332769800&to=1332845100&metric=uid&mode=json
		serverusage_api.php?from=1332769800&metric=uidt&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=ip&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=ipt&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=uip&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=uipt&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&mode=json
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&uid=320&mode=json

	CSV:
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&uid=320&mode=csv

	BASE64 ENCODED PHP SERIALIZED ARRAY:
		serverusage_api.php?from=1332769800&to=1332845100&metric=all&uid=320&mode=psa

	SVG:
		serverusage_api.php?from=1332769800&to=1332845100&mode=svg&width=1024&height=750&scale=log
		serverusage_api.php?from=1333532663&to=1333627917&mode=svg&scale=log&bgcol=light&gtype=12345
		serverusage_api.php?from=1333532663&to=1333627917&mode=svg&scale=linear&bgcol=light&gtype=15
		serverusage_api.php?from=1333532663&to=1333627917&mode=svg&scale=log&bgcol=light&gtype=5

OUTPUT:

	The output format can be JSON (JavaScript Object Notation), CSV (tab-separated text values), Base64 encoded serialized array or SVG (Scalable Vector Graphics).

ADDITIONAL NOTES:

	The reference time for all servers should be the standard UTC.
	The latest available time on the ServerUsage-Server aggregated table is always in the past by the value specified by DB_AGGREGATION_DELAY constant (by default 5 minutes).
	A valid time interval, for example, can be calculated as follows:

		polling_interval = 900; // 15 minutes * 60 seconds; must be equal or greater than DB_AGGREGATION_DELAY.
		delay_time = 600; // 10 minutes * 60 seconds; must be equal or greater than (2 * DB_AGGREGATION_DELAY).
		end_time = (current_time - delay_time);
		start_time = (end_time - polling_interval);

*/

//\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
// -----------------------------------------------------------------------------

// READ CONFIGURATION FILE

$conf = parse_ini_file('/etc/serverusage_server.conf');

/**
 * database name (serverusage)
 */
define ('K_DATABASE_NAME', $conf['SQLITE_DATABASE']);

// -----------------------------------------------------------------------------
//\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
// -----------------------------------------------------------------------------

// check input parameters

// get the starting time
if (isset($_GET['from'])) {
	$from_time = max(0, intval($_GET['from']));
} else {
	header('Missing \'from\' parameter', true, 501);
	exit;
}
// maximum end time that can be returned
$max_end_time = (time() - (2 * intval($conf['DB_AGGREGATION_DELAY'])));
// get the end time
if (isset($_GET['to'])) {
	$to_time = min($max_end_time, intval($_GET['to']));
} else {
	// set the default value
	$to_time = $max_end_time;
}
// check start time
if ($to_time < ($from_time + 60)) {
	// starting time must be greater than zero
	header('The specified time interval is too short', true, 501);
	exit;
}
if (!isset($_GET['mode'])) {
	// default JSON format output
	$mode = 'json';
} elseif (in_array($_GET['mode'], array('json', 'csv', 'psa', 'svg'))) {
	$mode = $_GET['mode'];
} else {
	header('Wrong \'mode\' parameter [\'json\', \'csv\', \'psa\', \'svg\']', true, 501);
	exit;
}
if ($mode == 'svg') {
	// force metric type for SVG output
	$metric = 'glb';
	if (isset($_GET['width'])) {
		$width = max(50, intval($_GET['width']));
	} else {
		// default width
		$width = 1024;
	}
	if (isset($_GET['height'])) {
		// round to a multiple of 5
		$height = (5 * intval(max(50, intval($_GET['height'])) / 5));
	} else {
		// default height
		$height = 750;
	}
	if (isset($_GET['scale']) AND ($_GET['scale'] == 'log')) {
		// print graph in logarithmic scale
		$logscale = true;
	} else {
		$logscale = false;
	}
	if (isset($_GET['bgcol']) AND ($_GET['bgcol']) == 'dark') {
		// dark background color
		$darkbg = true;
		$fcolor = 'ffffff';
		$bcolor = '000000';
	} else {
		$darkbg = false;
		$fcolor = '000000';
		$bcolor = 'ffffff';
	}
	if (isset($_GET['gtype']) AND (intval($_GET['gtype']) > 0)) {
		// select graphs to display
		$gseq = trim($_GET['gtype']);
		$numgraphs = strlen($gseq);
		$gtype = array();
		for ($i = 0; $i < $numgraphs; ++$i) {
			$gtype[] = ($gseq[$i] + 1);
		}
	} else {
		// print all available graphs
		$numgraphs = 5;
		$gtype = array(2,3,4,5,6);
	}
} elseif (isset($_GET['metric']) AND in_array($_GET['metric'], array('uid', 'uidt', 'ip', 'ipt', 'uip', 'uipt', 'grp', 'glb', 'all'))) {
	$metric = $_GET['metric'];
} else {
	header('Wrong \'metric\' parameter [\'uid\', \'uidt\', \'ip\', \'ipt\', \'uip\', \'uipt\', \'grp\', \'glb\', \'all\']', true, 501);
	exit;
}

// -----------------------------------------------------------------------------

// connect to database
try {
	$db = new PDO('sqlite:'.K_DATABASE_NAME);
} catch (PDOException $e) {
	header('SQLite connection error: '.$e->getMessage(), true, 503);
	exit;
}

// create the "where" portion of SQL query
$sqlwhere = ' WHERE lah_start_time>='.$from_time.' AND lah_end_time<='.$to_time.'';
if (isset($_GET['uid']) AND (($uid = intval($_GET['uid'])) >= 0)) {
	$sqlwhere .= ' AND lah_user_id='.$uid.'';
}
if (isset($_GET['ip']) AND (preg_match("/^([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})$/", $_GET['ip']) > 0)) {
	$sqlwhere .= ' AND lah_ip=\''.$_GET['ip'].'\'';
}

// select data by metric type
switch ($metric) {
	case 'uid' : {
		$sql = 'SELECT lah_user_id, SUM(lah_cpu_ticks) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_user_id';
		break;
	}
	case 'uidt' : {
		$sql = 'SELECT lah_user_id, SUM(lah_cpu_ticks), MIN(lah_start_time), MAX(lah_end_time) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_user_id';
		break;
	}
	case 'ip' : {
		$sql = 'SELECT lah_ip, SUM(lah_netin), SUM(lah_netout) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_ip';
		break;
	}
	case 'ipt' : {
		$sql = 'SELECT lah_ip, SUM(lah_netin), SUM(lah_netout), MIN(lah_start_time), MAX(lah_end_time) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_ip';
		break;
	}
	case 'uip' : {
		$sql = 'SELECT lah_user_id, lah_ip, SUM(lah_cpu_ticks), SUM(lah_io_read), SUM(lah_io_write), SUM(lah_netin), SUM(lah_netout) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_user_id, lah_ip';
		break;
	}
	case 'uipt' : {
		$sql = 'SELECT lah_user_id, lah_ip, SUM(lah_cpu_ticks), SUM(lah_io_read), SUM(lah_io_write), SUM(lah_netin), SUM(lah_netout), MIN(lah_start_time), MAX(lah_end_time) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_user_id, lah_ip';
		break;
	}
	case 'grp' : {
		$sql = 'SELECT lah_start_time, lah_end_time, lah_user_id, lah_ip, SUM(lah_cpu_ticks), SUM(lah_io_read), SUM(lah_io_write), SUM(lah_netin), SUM(lah_netout) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_start_time, lah_end_time, lah_user_id, lah_ip ORDER BY lah_end_time';
		break;
	}
	case 'glb' : {
		$sql = 'SELECT lah_start_time, lah_end_time, SUM(lah_cpu_ticks), SUM(lah_io_read), SUM(lah_io_write), SUM(lah_netin), SUM(lah_netout) FROM log_agg_hst'.$sqlwhere.' GROUP BY lah_start_time, lah_end_time ORDER BY lah_start_time';
		break;
	}
	case 'all' : {
		$sql = 'SELECT lah_start_time, lah_end_time, lah_process, lah_user_id, lah_ip, lah_cpu_ticks, lah_io_read, lah_io_write, lah_netin, lah_netout FROM log_agg_hst'.$sqlwhere.' ORDER BY lah_end_time';
		break;
	}
}

// extract requested data
$data = array();
try {
	$r = $db->query($sql, PDO::FETCH_NUM);
	foreach ($r as $row) {
		$data[] = $row;
	}
} catch (PDOException $e) {
	header('SQLite query error: '.$e->getMessage(), true, 503);
	exit;
}

// output data
switch ($mode) {
	case 'csv' : {
		// send headers
		header('Content-Description: TSV Data');
		header('Cache-Control: public, must-revalidate, max-age=0'); // HTTP/1.1
		header('Pragma: public');
		header('Expires: Sat, 26 Jul 1997 05:00:00 GMT'); // Date in the past
		header('Last-Modified: '.gmdate('D, d M Y H:i:s').' GMT');
		header('Content-Type: text/csv', false);
		// Turn on output buffering with the gzhandler
		ob_start('ob_gzhandler');
		// output code as TSV tab-separated values
		foreach ($data as $row) {
			echo implode("\t", $row)."\n";
		}
		break;
	}
	case 'psa' : {
		// send headers
		header('Content-Description: Serialized PHP array');
		header('Cache-Control: public, must-revalidate, max-age=0'); // HTTP/1.1
		header('Pragma: public');
		header('Expires: Sat, 26 Jul 1997 05:00:00 GMT'); // Date in the past
		header('Last-Modified: '.gmdate('D, d M Y H:i:s').' GMT');
		header('Content-Type: text/plain', false);
		// Turn on output buffering with the gzhandler
		ob_start('ob_gzhandler');
		// output code as base64 encoded PHP serialized array
		echo base64_encode(serialize($data));
		break;
	}
	case 'json' : {
		// send headers
		header('Content-Description: JSON Data');
		header('Cache-Control: public, must-revalidate, max-age=0'); // HTTP/1.1
		header('Pragma: public');
		header('Expires: Sat, 26 Jul 1997 05:00:00 GMT'); // Date in the past
		header('Last-Modified: '.gmdate('D, d M Y H:i:s').' GMT');
		header('Content-Type: application/json', false);
		header('Content-Transfer-Encoding: binary');
		// Turn on output buffering with the gzhandler
		ob_start('ob_gzhandler');
		// output code in JSON format
		echo json_encode($data);
		break;
	}
	case 'svg' : {
		// convert array elements to integers and calculate max values
		$maxval = array(0,0,0,0,0,0,0);
		foreach ($data as $k => $v) {
			$data[$k] = array_map('intval', $data[$k]);
			for ($i = 0; $i <=6; ++$i) {
				$maxval[$i] = max($maxval[$i], $data[$k][$i]);
			}
		}
		// colors for lines
		$color = array(2 => 'ab00dc', 3 => '184296', 4 => '589c00', 5 => 'dc7200', 6 => 'b71313');
		// colors for filling area
		$color_fill = array(2 => 'ce26ff', 3 => '2565e9', 4 => '75d000', 5 => 'ff9626', 6 => 'e92525');
		// labels
		$label = array(2 => 'CPU TICKS', 3 => 'IO READ', 4 => 'IO WRITE', 5 => 'NET IN', 6 => 'NET OUT');
		// units
		$unitlabel = array(2 => 'ticks', 3 => 'bytes', 4 => 'bytes', 5 => 'bytes', 6 => 'bytes');
		// horizontal ratio
		$hratio = ($width / ($maxval[1] - $data[0][0]));
		// height of each graph
		$graph_height = round($height / $numgraphs);
		// base for logarithm transformation
		$logbase = array(0,0);
		// vertical scale ratio
		$vratio = array(0,0);
		// initialize line coordinates
		$xs = 0;
		$xe = sprintf('%.3F', (($data[0][1] - $data[0][0]) * $hratio));
		$j = 0;
		for ($i = 2; $i <= 6; ++$i) {
			if (in_array($i, $gtype)) {
				if ($logscale) {
					// logarithm base
					$logbase[$i] = pow(($maxval[$i] + 1), (1.0 / $graph_height));
					// starting point
					$y = sprintf('%.3F', ($height - (($j * $graph_height) + log(($data[0][$i] + 1), $logbase[$i]))));
				} else {
					// vertical ratio
					$vratio[$i] = ($graph_height / $maxval[$i]);
					// starting point
					$y = sprintf('%.3F', ($height - (($j * $graph_height) + ($data[0][$i] * $vratio[$i]))));
				}
				// draw starting points for each column type
				$line[$i] = '<path fill="#'.$color_fill[$i].'" stroke="#'.$color[$i].'" stroke-width="1" d="M 0 '.($height - ($j * $graph_height)).' L '.$xs.' '.$y.' '.$xe.' '.$y.'';
				++$j;
			}
		}
		// create SVG graph
		$svg = '<'.'?'.'xml version="1.0" standalone="no"'.'?'.'>'."\n";
		$svg .= '<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">'."\n";
		$svg .= '<svg width="'.$width.'" height="'.$height.'" version="1.1" xmlns="http://www.w3.org/2000/svg">'."\n";
		$svg .= "\t".'<desc>SERVERUSAGE DATA FROM:'.$from_time.' TO:'.$to_time.'</desc>'."\n";
		// count the number of time points
		$numpoints = count($data);
		// draw lines between points
		for ($k = 1; $k < $numpoints; ++$k) {
			$xs = sprintf('%.3F', (($data[$k][0] - $data[0][0]) * $hratio));
			$xe = sprintf('%.3F', (($data[$k][1] - $data[0][0]) * $hratio));
			$j = 0;
			for ($i = 2; $i <= 6; ++$i) {
				if (in_array($i, $gtype)) {
					if ($logscale) {
						$y = sprintf('%.3F', ($height - (($j * $graph_height) + log(($data[$k][$i] + 1), $logbase[$i]))));
					} else {
						$y = sprintf('%.3F', ($height - (($j * $graph_height) + ($data[$k][$i] * $vratio[$i]))));
					}
					$line[$i] .= ' '.$xs.' '.$y.' '.$xe.' '.$y.'';
					++$j;
				}
			}
		}
		// close and print lines
		$j = 0;
		for ($i = 2; $i <= 6; ++$i) {
			if (in_array($i, $gtype)) {
				$svg .= $line[$i].' '.$width.' '.($height - ($j * $graph_height)).' z" />'."\n";
				++$j;
			}
		}
		// display vertical grid (time grid)
		$svg .= '<g stroke="#'.$fcolor.'" stroke-width="1">'."\n";
		for ($k = 1; $k < $numpoints; ++$k) {
			// one line every 10 minutes (600 seconds)
			if (($data[$k][0] % 600) == 0) {
				$x = sprintf('%.3F', (($data[$k][0] - $data[0][0]) * $hratio));
				$svg .= "\t".'<line x1="'.$x.'" y1="0" x2="'.$x.'" y2="'.$height.'"';
				if (($data[$k][0] % 3600) == 0) {
					// mark hour lines
					$svg .= ' opacity="0.3"';
				} else {
					$svg .= ' opacity="0.15"';
				}
				$svg .= ' />'."\n";
			}
		}
		$svg .= '</g>'."\n";
		// horizontal line grids
		$svg .= '<g stroke="#'.$fcolor.'" stroke-width="1">'."\n";
		$j = 0;
		for ($i = 2; $i <= 6; ++$i) {
			if (in_array($i, $gtype)) {
				// line distance
				$ld = pow(10, floor(log(($maxval[$i] / 2), 10)));
				// print lines
				for ($ln = $ld; $ln < $maxval[$i]; $ln += $ld) {
					if ($logscale) {
						// logarithmic
						$y = sprintf('%.3F', ($height - (($j * $graph_height) + log(($ln + 1), $logbase[$i]))));
					} else {
						// linear
						$y = sprintf('%.3F', ($height - (($j * $graph_height) + ($ln * $vratio[$i]))));
					}
					$svg .= "\t".'<line x1="0" y1="'.$y.'" x2="'.$width.'" y2="'.$y.'" opacity="0.3" />'."\n";
				}
				// print a label next to the first line
				$fontsize = round(min(($graph_height / 12),($width / 60)));
				if ($logscale) {
					// logarithmic
					$y = sprintf('%.3F', ($height - (($j * $graph_height) + log(($ld + 1), $logbase[$i]))));
					$x = 5;
				} else {
					// linear
					$y = sprintf('%.3F', ($height - (($j * $graph_height) + ($ld * $vratio[$i]))));
					$x = sprintf('%.3F', ($width / 2));
				}
				$svg .= '<text x="'.$x.'" y="'.($y + $fontsize).'" stroke-width="0" text-anchor="start" font-family="Arial,Verdana" font-size="'.$fontsize.'" fill="#'.$fcolor.'" opacity="0.4">'.number_format($ld,0,'.',',').'</text>'."\n";
				++$j;
			}
		}
		$svg .= '</g>'."\n";
		// add labels
		$svg .= '<g>'."\n";
		$x = ($width / 2);
		$j = 0;
		for ($i = 2; $i <= 6; ++$i) {
			if (in_array($i, $gtype)) {
				$y = ($height - ($j * $graph_height) - ($graph_height / 2));
				$fontsize = round(min(($graph_height / 6),($width / 30)));
				$svg .= '<text x="'.$x.'" y="'.$y.'" opacity="0.7" text-anchor="middle" font-family="Arial,Verdana" font-weight="bold" font-size="'.$fontsize.'" fill="#'.$fcolor.'">'.$label[$i].'</text>'."\n";
				$fontsize = round(min(($graph_height / 9),($width / 50)));
				$svg .= '<text x="'.$x.'" y="'.($y + $fontsize).'" opacity="0.7" text-anchor="middle" font-family="Arial,Verdana" font-size="'.$fontsize.'" fill="#'.$fcolor.'">(MAX: '.number_format($maxval[$i],0,'.',',').' '.$unitlabel[$i].')</text>'."\n";
				++$j;
			}
		}
		// display time
		$y = ($height - 5);
		$fontsize = round(min(($graph_height / 10),($width / 50)));
		$svg .= '<text x="5" y="'.$y.'" opacity="0.6" text-anchor="start" font-family="Arial,Verdana" font-size="'.$fontsize.'" fill="#'.$fcolor.'">'.date('Y-m-d H:i:s', $data[0][0]).'</text>'."\n";
		$svg .= '<text x="'.$width.'" y="'.$y.'" opacity="0.6" text-anchor="end" font-family="Arial,Verdana" font-size="'.$fontsize.'" fill="#'.$fcolor.'">'.date('Y-m-d H:i:s', $maxval[1]).'</text>'."\n";
		$svg .= '</g>'."\n";
		// close SVG graph
		$svg .= '</svg>'."\n";
		// send headers
		header('Content-Description: SVG Data');
		header('Cache-Control: public, must-revalidate, max-age=0'); // HTTP/1.1
		header('Pragma: public');
		header('Expires: Sat, 26 Jul 1997 05:00:00 GMT'); // Date in the past
		header('Last-Modified: '.gmdate('D, d M Y H:i:s').' GMT');
		header('Content-Type: image/svg+xml');
		header('Content-Disposition: inline; filename="srvusg.svg";');
		// Turn on output buffering with the gzhandler
		ob_start('ob_gzhandler');
		// output SVG code
		echo $svg;
		break;
	}
}

//=============================================================================+
// END OF FILE
//=============================================================================+
